### PR TITLE
Added extension of Google Optimize for azure pipelines

### PR DESCRIPTION
**Pull request by DeepakDahiya**

### CODE REVIEW

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/index.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
+import { Optimizeclient } from './operations/optimizeclient';
+import {IVariation , IExperiment} from './operations/IOptimize';
```
Comment (Ajay-MS) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/index.ts]: Spacing is not consistent.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/index.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        let action : string = param.getAction();
+        let filePath : string | null = param.getFilePath() ;
+
+        let taskOperation = new TaskOperation() ;
+
+        let experiment: IExperiment ;
+
+        if(filePath.endsWith('.json') ) {
+            let rawdata = fs.readFileSync(filePath);
+            experiment = JSON.parse(rawdata);
```
Comment (Ajay-MS) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/index.ts]: JSON.parse can throw an exception for invalid JSON

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/index.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            experiment.trafficCoverage = trafficCoverage ;
+            await taskOperation.updateTrafficCoverage(experiment ) ;
+        }
+
+        if(equalWeighting != null ) {
+            experiment.equalWeighting = equalWeighting ;
+            await taskOperation.updateEqualWeighting(experiment) ;
+        }
+
+        (action === "StopExperiment" ) ? (await taskOperation.stopExperiment(experiment ))
```
Comment (Ajay-MS) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/index.ts]: should we add switch ?

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/models/TaskParameter.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            this.experimentName = (tl.getInput('experimentName' , true));
+            this.experimentId = (tl.getInput('experimentName' , true));
+            this.action = tl.getInput('action' , true);
+            this.trafficCoverage = tl.getInput('trafficCoverage' , false);
+            this.equalWeighting = tl.getInput('equalWeighting' , false);
+            this.filePath = tl.getInput( 'jsonFile', false)
+        }
+        catch(error) {
+
+            throw new Error("Input gathering failed in parameter.ts");
```
Comment (Ajay-MS) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/models/TaskParameter.ts]: Error message should be localised.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/models/TaskParameter.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    private  experimentName: string ;
+    private  experimentId :string ;
+    private  action:string ;
+    private  trafficCoverage : string | null;
+    private  equalWeighting : string | null;
+    private  filePath: string ;
+
+    constructor() {
+        try{
+            this.endpoint = tl.getInput('googleEndpoint' , true);
```
Comment (Ajay-MS) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/models/TaskParameter.ts]: Second argument here means optional. Can you validate if we passed argument value correctly.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/models/TaskParameter.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    public getAction() : string{
+        return this.action ;
+    }
+
+    public getTrafficCoverage() : number | null {
+        let traffic = parseFloat(this.trafficCoverage)
+        if(Number.isNaN(traffic) || traffic <= 0 || traffic > 1){
+            throw tl.loc("")
```
Comment (Ajay-MS) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/models/TaskParameter.ts]: Throwing empty error message.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/JwtHandler.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    	return this.urlEscape(encoded);
+    }
+
+    public authorize(callback){
+        let self = this
+    	let now = parseInt(Date.now() / 1000, 10); // Google wants us to use seconds
+    	let cipher ;
+    	let	signatureInput ;
+    	let	signatureKey = this.privatekey ;
+        //let	signatureKey = this.privatekey ;
```
Comment (Ajay-MS) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/JwtHandler.ts]: Remove commented line.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/TaskOperation.ts [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+            experiment.status = "ENDED";
+            await this.oxclient.updateExperiment(experiment  , this.param);
+        }
+        else{
+            throw tl.loc("StopFailed");
+        }
+
+    }
+
+    public async updateExperiment(experiment:IExperiment ):Promise<void>{
```
Comment (Ajay-MS) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/TaskOperation.ts]: Should this method be startExperiment instead of update experiment.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/index.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        }
+        else {
+            experiment = {
+                "id" : param.getExperimentId()
+            }
+        }
+
+        if(trafficCoverage != null ) {
+            experiment.trafficCoverage = trafficCoverage ;
+            await taskOperation.updateTrafficCoverage(experiment ) ;
```
Comment (Ajay-MS) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/index.ts]: We are making multiple rest calls for updating each property of the experiment. Should we first update all properties in the experiment and a single call with update the experiment.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/optimizeclient.ts [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+let request = require('request');
+let qs = require('querystring');
+
+export class Optimizeclient{
+	private jwthandler:JwtHandler ;
+	constructor(){
+		this.jwthandler = new JwtHandler() ;
+	}
+	public async updateExperiment<T extends IExperiment>( experiment: T , param : TaskParameter) {
+        this.jwthandler.authorize(async function(err:any , token:any) {
```
Comment (Ajay-MS) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/optimizeclient.ts]: Spacing and indentation in file at multiple places.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/optimizeclient.ts [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+            let url = "https://www.googleapis.com/analytics/v3/management" ;
```
Comment (Ajay-MS) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/optimizeclient.ts]: Instead of hard coding can we fetch this URL.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/optimizeclient.ts [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+            var testclient = new restm.RestClient(userAgent, url, [bearerHandler]);
+            let restRes = await testclient.update(`accounts/${param.getAccountId()}/webproperties/${param.getWebPropertyId()}/profiles/${param.getProfileId()}/experiments/${param.getExperimentId()}`, experiment);
+            tl.debug(`function: 'updateExperiment'. response: '${JSON.stringify(restRes)}'`);
+
+            if (restRes.statusCode != 200) {
+                tl.debug(`Unable to update experiment with Id: '${experiment.id}'. Response:`);
+                tl.debug(JSON.stringify(restRes));
+                throw tl.loc("FailedToUpdatedExperiment", experiment.id);
+            }
+            console.log("experiment updated") ;
```
Comment (Ajay-MS) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/optimizeclient.ts]: Remove one of multiple console.log

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/optimizeclient.ts [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+            console.log(tl.loc("ExperimentWithIdUpdatedSuccessfully", experiment.id));
+        })
+    }
+
+    public async getExperiment<T extends IExperiment>(param: TaskParameter ): Promise<T> {
+        let url = "https://www.googleapis.com/analytics/v3/management" ;
```
Comment (Ajay-MS) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/optimizeclient.ts]: If we can fetch this URL, we should. Otherwise it should have been defined at only one place as a constant.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json [hunk 14]
```diff
@@ -0,0 +1,165 @@
+{
+   "id":"e7307da0-85f4-11e9-9dec-fdd4f9b653a3",
+   "name":"GoogleOptimize",
+   "friendlyName":"Google Optimize",
+   "description":"Azure devops extension for google optimize and content experiments",
+   "author":"Microsoft Corporation",
+   "helpMarkDown":"Replace with markdown to show in help",
```
Comment (Ajay-MS) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json]: Update helpmarkdown

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+   "category":"Utility",
+   "visibility":[
+      "Build",
+      "Release"
+   ],
+   "demands":[
```
Comment (Ajay-MS) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json]: Should extension be in public preview.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+         "required":true,
+         "helpMarkDown":"Enter the account Id which can be found in account settings inside admin tab"
+      },
+      {
+         "name":"webPropertyId",
+         "type":"string",
+         "label":"Web Property Id",
+         "defaultValue":"UA-XXXXX-Y",
+         "required":true,
+         "helpMarkDown":"Enter the web Property Id which can be found in property settings inside admin tab. Type- UA-XXXXX-Y"
```
Comment (kmkumaran) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json]: would giving a link be more useful, https://support.google.com/analytics/answer/1042508

Comment (kmkumaran) [reply Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json]: The comment to provide link will be applicable to other relevant optimize or analytics properties that you take as input below.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+      {
+         "name":"experimentName",
+         "type":"pickList",
+         "label":"Experiment Name",
+         "defaultValue":"",
+         "required":true,
+         "properties":{
+            "EditableOptions":true
+         },
+         "helpMarkDown":"Choose which experiment needs to be updated"
```
Comment (kmkumaran) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json]: Since this is editable, I presume user can also provide experiment which has not been created and not just that needs to be updated. Is that right?

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+            "PauseExperiment":"Pause an experiment",
+            "StopExperiment":"Stop an experiment"
+
+         },
+         "helpMarkDown":"Select the action to be performed on above experiment"
+      },
+      {
+         "name":"jsonFile",
+         "type":"filePath",
+         "label":"Json File",
```
Comment (kmkumaran) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json]: The Json file label seem too generic. Is this a experimentation config file? What is it called in Google Optimize?

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/index.ts [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+
+        let taskOperation = new TaskOperation() ;
+
+        let experiment: IExperiment ;
+
+        if(filePath.endsWith('.json') ) {
+            let rawdata = fs.readFileSync(filePath);
+            experiment = JSON.parse(rawdata);
+            if(param.getExperimentId() != experiment.id) {
+                throw tl.loc("IdMismatch") ;
```
Comment (kmkumaran) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/index.ts]: Should we throw or should we override?

File: Extensions/GoogleOptimize/Src/README.md [hunk 20]
```diff
@@ -0,0 +1,41 @@
+# Google Optimize
+Optimize allows you to test variants of web pages and see how they perform against an objective that you specify. Optimize monitors the results of your experiment and tells you which variant is the leader. To get started:
```
Comment (kmkumaran) [inline Extensions/GoogleOptimize/Src/README.md]: I assume the Readme contents are reviewed by PMs. If not, please get their input.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/JwtHandler.ts [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+        let self = this
+    	let now = parseInt(Date.now() / 1000, 10); // Google wants us to use seconds
+    	let cipher ;
+    	let	signatureInput ;
+    	let	signatureKey = this.privatekey ;
+        let	signature ;
+    	let	jwt ;
+    	this.authClaimSet.iat = now;
+    	this.authClaimSet.exp = now + 60; // Token valid for one minute
+        signatureKey = signatureKey.replace(/\\n/g, '\n');
```
Comment (kmkumaran) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/JwtHandler.ts]: AuthClaimSet is being set, spread across constructor and here. Instead can you have a single private method to getAuthClaimSet?

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/JwtHandler.ts [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+    	cipher.update(signatureInput);
+    	signature = cipher.sign(signatureKey, 'base64');
+    	jwt = signatureInput + '.' + this.urlEscape(signature);
+    	// Send request to authorize this application
+    	request({
+    		method: 'POST',
+    		headers: {
+    			'Content-Type': 'application/x-www-form-urlencoded'
+    		},
+    		uri: 'https://accounts.google.com/o/oauth2/token',
```
Comment (kmkumaran) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/JwtHandler.ts]: Do you have this Uri in endpoint?

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/TaskOperation.ts [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+        }
+    }
+
+    public async stopExperiment(experiment:IExperiment ):Promise<void> {
+        if(this.currentExperiment.status !== "ENDED"){
+            experiment.status = "ENDED";
+            await this.oxclient.updateExperiment(experiment  , this.param);
+        }
+        else{
+            throw tl.loc("StopFailed");
```
Comment (kmkumaran) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/TaskOperation.ts]: The operations should be idempotent, if you attempt to stop a already stopped experiment, it shouldn't throw.
The requirement of operations being idempotent is true for all actions.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/package-lock.json [hunk 24]
```diff
@@ -0,0 +1,709 @@
+{
+  "name": "googleoptimize",
```
Comment (kmkumaran) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/package-lock.json]: Please add a Third Party notice for the modules that you are taking dependency on. @SumiranAgg  can help you on this.

File: Extensions/GoogleOptimize/Src/vss-extension.json [hunk 25]
```diff
@@ -0,0 +1,83 @@
+{
+   "manifestVersion":1,
+   "id":"GoogleOptimize",
+   "version":"1.0.0",
+   "publisher":"ms-vscs-rm",
+   "name":"Google Optimize",
+   "description":"Azure Devops extension for Google Optimizely and content Experiments integration.",
+   "public":false,
+   "icons":{
+      "default":"Tasks/GoogleOptimize/icon.png"
```
Comment (SumiranAgg) [inline Extensions/GoogleOptimize/Src/vss-extension.json]: Please verify icon with PMs.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json [hunk 26]
```diff
@@ -0,0 +1,166 @@
+{
+   "id":"e7307da0-85f4-11e9-9dec-fdd4f9b653a3",
+   "name":"GoogleOptimize",
+   "friendlyName":"Google Optimize",
+   "description":"Azure devops extension for google optimize and content experiments",
+   "author":"Microsoft Corporation",
+   "helpMarkDown":"[More Information](https://github.com/Microsoft/azure-pipelines-extensions/blob/master/Extensions/GoogleOptimize/Src/README.md)",
```
Comment (SumiranAgg) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json]: [Learn more about this task] instead of [More Information]

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json (hunk 17 above)
Comment (SumiranAgg) [reply Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json]: +1 
Should this be editable or should the user be allowed to pick only available experiments?

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json [hunk 27]
```diff
@@ -0,0 +1,166 @@
...
+         "helpMarkDown":"Choose which experiment needs to be updated"
+      },
+      {
+         "name":"action",
+         "type":"pickList",
+         "label":"Action",
+         "defaultValue":"",
+         "required":true,
+         "options":{
+            "UpdateExperiment":"Start/Update an experiment",
```
Comment (SumiranAgg) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json]: As per Roopesh's comment change action to: Start or Update experiment, Pause experiment, Stop experiment

**Comment by SumiranAgg** [comment]
Add a third party notice file

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/package.json [hunk 28]
```diff
@@ -0,0 +1,23 @@
+{
+  "name": "googleoptimize",
+  "version": "1.0.0",
+  "description": "",
```
Comment (SumiranAgg) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/package.json]: add description

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/models/TaskParameter.ts [hunk 29]
```diff
@@ -0,0 +1,74 @@
+import * as tl from 'azure-pipelines-task-lib/task';
+
+export class TaskParameter{
```
Comment (SumiranAgg) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/models/TaskParameter.ts]: Make it a sungleton class instead and use setters and getters for properties.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/optimizeclient.ts [hunk 30]
```diff
@@ -0,0 +1,52 @@
...
+                throw tl.loc("FailedToUpdatedExperiment", experiment.id);
+            }
+            console.log(tl.loc("ExperimentWithIdUpdatedSuccessfully", experiment.id));
+        })
+    }
+
+    public async getExperiment<T extends IExperiment>(param: TaskParameter ): Promise<T> {
+        this.jwthandler.authorize(async function(err:any, token:any) {
+            const bearerHandler = new hm.BearerCredentialHandler(token);
+            var testclient = new restm.RestClient(userAgent, url, [bearerHandler]);
```
Comment (SumiranAgg) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/optimizeclient.ts]: provide meaningful name to 'testclient'

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json (hunk 17 above)
Comment (DeepakDahiya) [reply Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json]: I changed it to false. User shouldn't be allowed to add anything other than drop down options.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json (hunk 18 above)
Comment (DeepakDahiya) [reply Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json]: Finalized "Experiment Configuration File" as the label

File: Extensions/GoogleOptimize/Src/.gitattributes [hunk 31]
```diff
@@ -0,0 +1,2 @@
+# Auto detect text files and perform LF normalization
+* text=auto
```
Comment (SumiranAgg) [inline Extensions/GoogleOptimize/Src/.gitattributes]: Do we need this file/ attribute?

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json (hunk 17 above)
Comment (SumiranAgg) [reply Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json]: Remove the properties. It is false by default.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json [hunk 32]
```diff
@@ -0,0 +1,185 @@
...
+         "helpMarkDown":"Web property id refers to your specific website on which experiment is set up"
+      },
+      {
+         "name":"profileId",
+         "type":"pickList",
+         "label":"Profile Id",
+         "defaultValue":"",
+         "required":true,
+         "properties":{
+            "EditableOptions":false
```
Comment (SumiranAgg) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json]: same as below

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json [hunk 33]
```diff
@@ -0,0 +1,185 @@
...
+      },
+      {
+         "name":"action",
+         "type":"pickList",
+         "label":"Action",
+         "defaultValue":"",
+         "required":true,
+         "options":{
+            "UpdateExperiment":"Start or Update experiment",
+            "StopExperiment":"Stop an experiment"
```
Comment (SumiranAgg) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/task.json]: remove "an"

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/models/TaskParameter.ts [hunk 34]
```diff
@@ -0,0 +1,119 @@
+import * as tl from 'azure-pipelines-task-lib/task';
```
Comment (SumiranAgg) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/models/TaskParameter.ts]: Fix indentation in this file

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/models/TaskParameter.ts [hunk 35]
```diff
@@ -0,0 +1,119 @@
...
+	get action(): string {
+		return this._action;
+	}
+
+	set action(value: string) {
+		this._action = value;
+	}
+
+	get trafficCoverage(): number | null {
+		let traffic = parseFloat(this._trafficCoverage);
```
Comment (SumiranAgg) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/models/TaskParameter.ts]: Should we move this validation when we are getting input? Which will allow us to fail early. Otherwise we fail only when a get is done on this property.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/models/TaskParameter.ts [hunk 36]
```diff
@@ -0,0 +1,119 @@
...
+	private _profileId: string;
+	private _experimentId: string;
+	private _action: string;
+	private _trafficCoverage: string | null;
+	private _equalWeighting: string | null;
+	private _filePath: string | null;
+
+	private static _taskParameters: TaskParameter = null;
+
+	constructor () {
```
Comment (SumiranAgg) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/models/TaskParameter.ts]: Make constructor private so a new instance cannot be created from outside the class and getInstance() is the only way to get an instance of this class.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/index.ts [hunk 37]
```diff
@@ -0,0 +1,79 @@
...
+            try {
+                experiment = JSON.parse(rawdata);
+            } catch (err) {
+                throw tl.loc("FailedToParseFile", err);
+            }
+
+            if (param.experimentId != experiment.id) {
+                throw tl.loc("IdMismatch");
+            }
+            for (var key in experiment) {
```
Comment (SumiranAgg) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/index.ts]: Nested properties are not validated here. Do we need to add logic for the same?

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/index.ts [hunk 38]
```diff
@@ -0,0 +1,79 @@
...
+                await taskOperation.stopExperiment(experiment);
+                break;
+            case "UpdateExperiment":
+                await taskOperation.updateExperiment(experiment);
+                break;
+            default:
+                throwError();
+        }
+    } catch (err) {
+        console.log(err);
```
Comment (SumiranAgg) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/index.ts]: Remove console.log.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/index.ts [hunk 39]
```diff
@@ -0,0 +1,79 @@
...
+        tl.setResult(tl.TaskResult.Failed, err);
+    }
+}
+
+function throwError() {
```
Comment (SumiranAgg) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/index.ts]: Why do you need this function? Can we use directly  throw tl.loc()

File: Extensions/GoogleOptimize/Src/.gitattributes (hunk 31 above)
Comment (DeepakDahiya) [reply Extensions/GoogleOptimize/Src/.gitattributes]: I'll remove this file.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/models/GoogleOptimizeUtils.ts [hunk 40]
```diff
@@ -0,0 +1,95 @@
...
+function urlEscape(source) {
+    return source.replace(/\+/g, '-').replace(/\//g, '_').replace(/\=+$/, '');
+}
+
+function base64Encode(obj) {
+    var encoded = new Buffer(JSON.stringify(obj), 'utf8').toString('base64');
+    return urlEscape(encoded);
+}
+
+export async function UpdateExperimentUtil(testclient: restm.RestClient, experiment: IExperiment) {
```
Comment (SumiranAgg) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/models/GoogleOptimizeUtils.ts]: Can we change testClient to something more meaningful

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/optimizeclient.ts [hunk 41]
```diff
@@ -0,0 +1,61 @@
...
+
+			let currentExperiment = await GetExperimentUtil(restclient);
+
+			if (currentExperiment.statusCode != 200) {
+				console.log(tl.loc("FailedToFetchCurrentExperiment", JSON.stringify(currentExperiment.error)));
+				tl.setResult(tl.TaskResult.Failed, err);
+				throw tl.loc("FailedToFetchCurrentExperiment" ,currentExperiment.error );
+			}
+
+			if(currentExperiment.result.status === "ENDED"){
```
Comment (SumiranAgg) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/optimizeclient.ts]: What if customer wanted to end the experiment but it was already stopped? This will fail the task.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/TaskOperation.ts [hunk 42]
```diff
@@ -0,0 +1,25 @@
...
+export class TaskOperation {
+
+	private oxclient: Optimizeclient;
+
+	constructor() {
+		this.oxclient = new Optimizeclient();
+	}
+
+    public async stopExperiment(experiment: IExperiment): Promise < void > {
+		experiment.status = "ENDED";
```
Comment (SumiranAgg) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/TaskOperation.ts]: Use string enum for statuses (Ended, Running). So that if string changes there is only one touch point.

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/optimizeclient.ts (hunk 41 above)
Comment (SumiranAgg) [reply Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/optimizeclient.ts]: Replace ENDED with enum val defined in taskoperation file

File: Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/TaskOperation.ts [hunk 43]
```diff
@@ -0,0 +1,25 @@
+import {IVariation,IExperiment} from './../models/IOptimize';
```
Comment (SumiranAgg) [inline Extensions/GoogleOptimize/Src/Tasks/GoogleOptimize/operations/TaskOperation.ts]: Remove IVariation import if not using in file. Check and remove unused imports in all file please.

File: package.json [hunk 44]
```diff
@@ -32,5 +32,8 @@
     "vsts-task-lib": "^2.0.5",
     "xml2js": "^0.4.17",
     "yargs": "^4.7.1"
+  },
+  "dependencies": {
+    "natives": "^1.1.6"
```
Comment (SumiranAgg) [inline package.json]: check if required?

File: package-lock.json [hunk 45]
```diff
@@ -5345,10 +5345,9 @@
       }
     },
     "natives": {
-      "version": "1.1.0",
-      "resolved": "https://registry.npmjs.org/natives/-/natives-1.1.0.tgz",
-      "integrity": "sha1-6f+EFBimsux6SV6TmYT3jxY+bjE=",
-      "dev": true
+      "version": "1.1.6",
+      "resolved": "https://registry.npmjs.org/natives/-/natives-1.1.6.tgz",
+      "integrity": "sha512-6+TDFewD4yxY14ptjKaS63GVdtKiES1pTPyxn9Jb0rBqPMZ7VcCiooEhPNsr+mqHtMGxa/5c/HhcC4uPEUw/nA=="
```
Comment (SumiranAgg) [inline package-lock.json]: Is this change required?
